GH-45096: [C++] Apply a cstdint patch to bundled Thrift for GCC 15

kou opened the pull request:
### Rationale for this change

Apache Thrift misses `#include <cstdint>` for `int64_t`: https://github.com/apache/thrift/pull/3078
GCC 15 requires it.

### What changes are included in this PR?

Apply https://github.com/apache/thrift/pull/3078 only with GCC 15.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* GitHub Issue: #45096

raulcd reviewed (approved):
This seems good to me. @jonkeane do you want to validate it?

jonkeane reviewed (approved):
Thanks for this, yeah this looks right, adding https://gcc.gnu.org/gcc-15/porting_to.html in case it is helpful for others. 

I can pick up trying to add this to CI — @kou do you have thoughts on getting `gcc-snapshot` installed reliably? I can keep on trying on #44970 but you'll know what I've been missing there.

kou commented:
> Thanks for this, yeah this looks right, adding https://gcc.gnu.org/gcc-15/porting_to.html in case it is helpful for others.

Thanks. I didn't notice this page. I've added the information to the Thrift PR.
 
> I can pick up trying to add this to CI — @kou do you have thoughts on getting `gcc-snapshot` installed reliably? I can keep on trying on #44970 but you'll know what I've been missing there.

I've added some comments to #44970.

pitrou commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1777,13 +1777,28 @@ macro(build_thrift)
   endif()
 
+  set(THRIFT_PATCH_COMMAND)
+  if(CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 15.0)
```
Why make this conditional? The patch should always be valid anyway.

raulcd replied:
that is fair, this has been merged without any conditional: https://github.com/apache/thrift/commit/947ad66940cfbadd9b24ba31d892dfc1142dd330

kou replied:
If we always use `PATCH_COMMAND`, `patch` command is required in a build machine. We want to avoid it. For example, `patch` command doesn't exist in most Windows machine.

pitrou replied:
Ah, right. But the condition seems overly specific in any case.

kou replied:
Hmm. I don't think that the current condition has a problem because GCC 15 is the only known problematic compiler for now. And this is a temporary condition. We will remove this patch when a new Apache Thrift is released.

What is your real concern?

pitrou replied:
I was just surprised, no specific concern. Especially if Thrift has incorporated the patch anyway.
